geotiepoints/modisinterpolator.py

```diff
@@ -78,83 +78,138 @@ def compute_expansion_alignment(satz_a, satz_b, satz_c, satz_d):
+
+    """
+    def __init__(self, coarse_resolution, fine_resolution, coarse_scan_width=None):
+        if coarse_resolution == 1000:
+            coarse_scan_length = 10
+            coarse_scan_width = 1354
+            self._get_coords = self._get_coords_1km
+            self._expand_tiepoint_array = self._expand_tiepoint_array_1km
+        elif coarse_resolution == 5000:
+            coarse_scan_length = 2
+            self._get_coords = self._get_coords_5km
+            self._expand_tiepoint_array = self._expand_tiepoint_array_5km
+            if coarse_scan_width is None:
+                coarse_scan_width = 271
             else:
-                self.cscan_full_width = cscan_full_width
-            self.expand_tiepoint_array = self._expand_tiepoint_array_5km
-            self.get_coords = self._get_coords_5km
-
-        if fres == 250:
-            self.fscan_width = 4 * self.cscan_width
-            self.fscan_full_width = 1354 * 4
-            self.fscan_len = 4 * 10 // self.cscan_len
-        elif fres == 500:
-            self.fscan_width = 2 * self.cscan_width
-            self.fscan_full_width = 1354 * 2
-            self.fscan_len = 2 * 10 // self.cscan_len
-        elif fres == 1000:
-            self.fscan_width = 1 * self.cscan_width
-            self.fscan_full_width = 1354
-            self.fscan_len = 1 * 10 // self.cscan_len
-
-    def _expand_tiepoint_array_1km(self, arr, lines, cols):
-        arr = da.repeat(arr, lines, axis=1)
-        arr = da.concatenate((arr[:, :lines//2, :], arr, arr[:, -(lines//2):, :]), axis=1)
-        arr = da.repeat(arr.reshape((-1, self.cscan_full_width - 1)), cols, axis=1)
-        return da.hstack((arr, arr[:, -cols:]))
+                coarse_scan_width = coarse_scan_width
+        self._coarse_scan_length = coarse_scan_length
+        self._coarse_scan_width = coarse_scan_width
+        self._coarse_pixels_per_1km = coarse_resolution // 1000
+
+        fine_pixels_per_1km = {
+            250: 4,
+            500: 2,
+            1000: 1,
+        }[fine_resolution]
+        self._fine_pixels_per_coarse_pixel = fine_pixels_per_1km * self._coarse_pixels_per_1km
+        self._fine_scan_width = 1354 * fine_pixels_per_1km
+        self._fine_scan_length = fine_pixels_per_1km * 10 // coarse_scan_length
+        self._coarse_resolution = coarse_resolution
+        self._fine_resolution = fine_resolution
+
+    def interpolate(self, lon1, lat1, satz1):
+        return self._interpolate(lon1, lat1, satz1,
+                                 coarse_resolution=self._coarse_resolution,
+                                 fine_resolution=self._fine_resolution,
+                                 coarse_scan_width=self._coarse_scan_width)
+    @scanline_mapblocks
+    def _interpolate(
+        self,
+        lon1,
+        lat1,
+        satz1,
+        coarse_resolution=None,
+        fine_resolution=None,
+        coarse_scan_width=None,
```

Comment:
why do these need to be passed vs retrieve from `self`?

Reply:
Read the note in the docstring. The decorator uses them to determine what the output shape and chunk size is going to be.

Reply:
that's a bit ugly :)

Reply:
I could move the `__init__` logic into this method but then the `__init__` has no purpose. I could change this back into a function (again) and create the class inside the map_blocks'd function. Then this all makes sense.